source/code/omi_preexec/mycnf-scanner.cpp

```diff
@@ -0,0 +1,251 @@
+                    portNum = SCXCoreLib::StrToUInt( elements[1] );
+                }
+
+                // Check for a bind-address declaration (bind-address = 172.0.0.1)
+                if ( 0 == SCXCoreLib::StrCompare(L"bind-address", elements[0], true) )
+                {
+                    SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L"   Found bind-address: ", elements[1]));
+                    bindAddress = elements[1];
+                }
+            }
+
+            if ( portNum )
+            {
+                SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(SCXCoreLib::StrAppend(L"   Flushing port ", portNum).append(L", Bind Address: "), bindAddress));
+
+                auth.AddCredentialSet(portNum, SCXCoreLib::StrToUTF8(bindAddress));
+                m_ports[portNum] = bindAddress;
+            }
+        }
+    }
+
+    /**
+       call LoadCnfFile on each ".cnf" file present in the given directory.
+
+       \param[in]  directory                     path of the MySQL configuration file
+       \param[in]  auth                          authentication object that has been 'Load'ed.
+       \throws     SCXFilePathNotFoundException  if directory is not found
+    */
+    void MySQL_ConfigurationScanner:: LoadCnfDirectory(
+        const SCXCoreLib::SCXFilePath& directory, 
+        MySQL_Authentication& auth)
+    {
+        SCX_LOGHYSTERICAL(m_log, SCXCoreLib::StrAppend(L"LoadCnfFilesFromDirectory", directory.Get()));
+        std::vector< SCXCoreLib::SCXFilePath > cnfFiles = m_extDepends->GetFiles(directory);
+
+        for( std::vector< SCXCoreLib::SCXFilePath >::const_iterator iter = cnfFiles.begin(); iter != cnfFiles.end(); ++iter)
+        {
+            std::wstring fileSuffix = iter->GetFilesuffix();
+            if (fileSuffix == L"cnf")
+            {
+                LoadCnfFile(*iter, auth);
+            }
+            else
+            {
+                SCX_LOGHYSTERICAL(m_log,SCXCoreLib::StrAppend(iter->Get(), L" is not a cnf file"));
+            }
+        }
+    }
+
+    /**
+       Parse a MySQL configuration file looking for 'port' and 'bind-address'
+       options, and add an entry to the given auth file for each 'port'/
+       'bind-address' pair.
+       '!include' and '!includedir' directives are honored.
+
+       \param[in]  cnfFile                       path of the MySQL configuration file
+       \param[in]  auth                          authentication object that has been 'Load'ed.
+       \throws     SCXFilePathNotFoundException  if cnfFile (or any of the paths pointed by "!include" directives) is not found
+       \throws     MySQL_ConfigScannerException  if the include directives form a cycle
+    */
+    void MySQL_ConfigurationScanner::Scan(const SCXCoreLib::SCXFilePath& cnfFile, MySQL_Authentication& auth)
+    {
+        m_visited.clear();
+        m_ports.clear();
+        LoadCnfFile(cnfFile, auth);
+
+        std::vector<unsigned int> authPorts;
```

Comment:
These comments came from original code, and it's worth including here (took me a minute to understand what was happening here):

// We've added what must be added, but we need to delete unreferenced ports
// (Loop through authentication ports and remove anything we didn't read from configuration